### PR TITLE
ILI-28 Added an ORM_O01 message as an attachment to DocumentSetRequestType

### PR DESCRIPTION
**Pull request by wojtekbu**
* Refactored xdssender.MessageUtil (splited method into parts)
* Moved buildDocInfo() method from XdsExportServiceImpl to DocumentInfo (as a constructor)

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/xdssender/api/cda/model/DocumentModel.java [hunk 1]
```diff
@@ -34,12 +36,15 @@
...
-		
 	}
 	
 	public ClinicalDocument getDoc() {
 		return doc;
 	}
+
+    public List<Author> getAuthors() {
+        return doc != null ? doc.getAuthor() : new ArrayList<Author>();
+    }
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/xdssender/api/cda/model/DocumentModel.java]: There is something wrong with indentation. Tip: enable displaying white characters -> https://intellij-support.jetbrains.com/hc/en-us/community/posts/207045165-Show-non-printable-characters-Tabs-

File: api/src/main/java/org/openmrs/module/xdssender/api/model/DocumentInfo.java [hunk 2]
```diff
@@ -49,6 +55,27 @@
...
+		formatCode = docModel.getFormatCode();
+		creationTime = new Date();
+		this.mimeType = mimeType;
+		this.patient = patient;
+
+		List<Provider> provs = new ArrayList<Provider>();
+		for (Author aut : docModel.getAuthors()) {
+			// Load the author
+			for (II id : aut.getAssignedAuthor().getId())
+				if (id.getRoot() != null && id.getRoot().equals(providerRoot))
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/xdssender/api/model/DocumentInfo.java]: I think you should you StringUtils here

File: api/src/main/java/org/openmrs/module/xdssender/api/service/impl/XdsExportServiceImpl.java [hunk 3]
```diff
@@ -42,49 +42,34 @@
...
 				throw new Exception("Could not execute provide and register");
-			
-			return docInfo;
+
+			return clinicalDocInfo;
 		}
 		catch (Exception e) {
 			throw new RuntimeException(e);
 		}
 	}
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/xdssender/api/service/impl/XdsExportServiceImpl.java]: Something wrong with the indentation in the method above

File: api/src/main/java/org/openmrs/module/xdssender/api/service/impl/XdsExportServiceImpl.java [hunk 4]
```diff
@@ -42,49 +42,34 @@
 	@Override
 	public DocumentInfo exportProvideAndRegister(Encounter encounter, Patient patient) {
 		try {
-			DocumentModel docModel = docBuilder.buildDocument(patient, encounter);
-			
-			DocumentInfo docInfo = buildDocInfo(encounter, patient, docModel);
-			
-			if (!messageUtil.getPatientIdentifier(docInfo).getIdentifierType().getName().equals("ECID")) {
+			DocumentModel clinicalDocModel = docBuilder.buildDocument(patient, encounter);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/xdssender/api/service/impl/XdsExportServiceImpl.java]: I think changing the name docBuilder would be a good idea here.

File: pom.xml [hunk 5]
```diff
@@ -41,6 +41,18 @@
...
+                <artifactId>labintegration-api</artifactId>
+                <version>1.0.0-SNAPSHOT</version>
+                <scope>provided</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.openmrs.web</groupId>
+                <artifactId>openmrs-web</artifactId>
+                <version>${openmrsPlatformVersion}</version>
+                <scope>provided</scope>
+            </dependency>
```
Comment (dserkowski) [inline pom.xml]: indentation

File: omod/src/main/resources/config.xml [hunk 6]
```diff
@@ -18,6 +18,9 @@
 		<require_module version="${eventVersion}">
 			org.openmrs.event
 		</require_module>
+		<require_module>
+            org.openmrs.module.labintegration
```
Comment (dserkowski) [inline omod/src/main/resources/config.xml]: indentation

File: pom.xml [hunk 7]
```diff
@@ -41,6 +41,18 @@
...
+                <groupId>org.openmrs.module</groupId>
+                <artifactId>labintegration-api</artifactId>
+                <version>1.0.0-SNAPSHOT</version>
+                <scope>provided</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.openmrs.web</groupId>
+                <artifactId>openmrs-web</artifactId>
+                <version>${openmrsPlatformVersion}</version>
+                <scope>provided</scope>
```
Comment (dserkowski) [inline pom.xml]: Do you need org.openmrs.web?

File: api/src/main/java/org/openmrs/module/xdssender/api/cda/ORM_O01DocumentBuilder.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	private final Log log = LogFactory.getLog(this.getClass());
+
+	private static final String ORM_O01 = "ORM_O01";
+
+    @Autowired
+    private OrderGeneratorManager orderGeneratorManager;
+	
+	public DocumentModel buildDocument(Encounter encounter) {
+        try {
+            String message = orderGeneratorManager.generateORMO01Message(encounter, OrderDestination.SCC);
```
Comment (dserkowski) [inline api/src/main/java/org/openmrs/module/xdssender/api/cda/ORM_O01DocumentBuilder.java]: Shouldn't OrderDestination be a parameter of the buildDocument method?

File: api/src/main/java/org/openmrs/module/xdssender/api/cda/ClinicalDocumentBuilder.java [hunk 9]
```diff
@@ -108,7 +106,8 @@ public DocumentModel buildDocument(Patient patient, Encounter encounter) throws
 			XmlIts1Formatter formatter = EverestUtil.createFormatter();
 			formatter.graph(baos, doc);
 
-			return DocumentModel.createInstance(baos.toByteArray(), builder.getTypeCode(), builder.getFormatCode(), doc);
+			return DocumentModel.createInstance(baos.toByteArray(), builder.getTypeCode(),
+                XdsSenderConstants.CODE_SYSTEM_NAME_LOINC, builder.getFormatCode(), doc);
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/xdssender/api/cda/ClinicalDocumentBuilder.java]: fix ident

File: api/src/main/java/org/openmrs/module/xdssender/api/xds/ExtrinsicObject.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                lastEncounterTs.getValue());
+
+    }
+
+    public void addSourcePatientInformation(DocumentInfo info, PatientIdentifier nationalIdentifier,
+                                            PatientIdentifier siteIdentifier, XdsSenderConfig config,
+                                            TS patientDob) throws JAXBException {
+        InfosetUtil.addOrOverwriteSlot(this, XDSConstants.SLOT_NAME_SOURCE_PATIENT_ID,
+                String.format("%s^^^&%s&ISO", patientId, config.getEcidRoot()));
+        InfosetUtil.addOrOverwriteSlot(this, XDSConstants.SLOT_NAME_SOURCE_PATIENT_INFO,
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/xdssender/api/xds/ExtrinsicObject.java]: fix ident

File: api/src/main/java/org/openmrs/module/xdssender/api/xds/ExtrinsicObject.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        InfosetUtil.addOrOverwriteSlot(this, XDSConstants.SLOT_NAME_SERVICE_STOP_TIME,
+                lastEncounterTs.getValue());
+
+    }
+
+    public void addSourcePatientInformation(DocumentInfo info, PatientIdentifier nationalIdentifier,
+                                            PatientIdentifier siteIdentifier, XdsSenderConfig config,
+                                            TS patientDob) throws JAXBException {
+        InfosetUtil.addOrOverwriteSlot(this, XDSConstants.SLOT_NAME_SOURCE_PATIENT_ID,
+                String.format("%s^^^&%s&ISO", patientId, config.getEcidRoot()));
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/xdssender/api/xds/ExtrinsicObject.java]: fix ident

File: api/src/main/java/org/openmrs/module/xdssender/api/xds/MessageUtil.java [hunk 12]
```diff
@@ -125,8 +139,151 @@ private ProvideAndRegisterDocumentSetRequestType addDocToRequest(ProvideAndRegis
         return requestType;
     }
 
-	private SubmitObjectsRequest createClinicalSubmitRequest(DocumentInfo info,
-                                Encounter encounter, ExtrinsicObjectType oddRegistryObject) throws JAXBException  {
+    private SubmitObjectsRequest addToRegistryRequest(SubmitObjectsRequest registryRequest, DocumentData docData,
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/xdssender/api/xds/MessageUtil.java]: why did you have add so much code? What does it do exactly?

Comment (wojtekbu) [reply api/src/main/java/org/openmrs/module/xdssender/api/xds/MessageUtil.java]: I've:
* Splited old createProvideAndRegisterDocument method into:
createClinicalSubmitRequest and createClinicalExtrinsicObjectType
(logic responsible for generating CDA stays the same)

* Refactored logic of createProvideAndRegisterDocument by spliting it's
logic into several methods (used for generating the HL7 attachment)

createClinicalSubmitRequest and createClinicalExtrinsicObjectType can be relaplaced with new metody aftherwards

**Pull request closed without merging**